cmd/query/app/querysvc/adjuster/clockskew.go

```diff
@@ -0,0 +1,201 @@
+// or points to an ID for which there is no span.
+func (a *clockSkewAdjuster) buildSubGraphs() {
+	a.roots = make(map[pcommon.SpanID]*node)
+	for _, n := range a.spans {
+		if n.span.ParentSpanID() == pcommon.NewSpanIDEmpty() {
+			a.roots[n.span.SpanID()] = n
+			continue
+		}
+		if p, ok := a.spans[n.span.ParentSpanID()]; ok {
+			p.children = append(p.children, n)
+		} else {
+			warning := fmt.Sprintf(warningFormatInvalidParentID, n.span.ParentSpanID())
+			jptrace.AddWarning(n.span, warning)
+			// treat spans with invalid parent ID as root spans
+			a.roots[n.span.SpanID()] = n
+		}
+	}
+}
+
+func (a *clockSkewAdjuster) adjustNode(n *node, parent *node, skew clockSkew) {
+	if (n.hostKey != skew.hostKey || n.hostKey == "") && parent != nil {
+		// Node n is from a different host. The parent has already been adjusted,
+		// so we can compare this node's timestamps against the parent.
+		skew = clockSkew{
+			hostKey: n.hostKey,
+			delta:   a.calculateSkew(n, parent),
+		}
+	}
+	a.adjustTimestamps(n, skew)
+	for _, child := range n.children {
+		a.adjustNode(child, n, skew)
+	}
+}
+
+func (*clockSkewAdjuster) calculateSkew(child *node, parent *node) time.Duration {
+	parentStartTime := parent.span.StartTimestamp().AsTime()
+	childStartTime := child.span.StartTimestamp().AsTime()
+	parentEndTime := parent.span.EndTimestamp().AsTime()
+	childEndTime := child.span.EndTimestamp().AsTime()
+	parentDuration := parentEndTime.Sub(parentStartTime)
+	childDuration := childEndTime.Sub(childStartTime)
+
+	if childDuration > parentDuration {
+		// When the child lasted longer than the parent, it was either
+		// async or the parent may have timed out before child responded.
+		// The only reasonable adjustment we can do in this case is to make
+		// sure the child does not start before parent.
+		if childStartTime.Before(parentStartTime) {
+			return parentStartTime.Sub(childStartTime)
+		}
+		return 0
+	}
+	if !childStartTime.Before(parentStartTime) && !childEndTime.After(parentEndTime) {
+		// child already fits within the parent span, do not adjust
+		return 0
+	}
+	// Assume that network latency is equally split between req and res.
+	latency := (parentDuration - childDuration) / 2
+	// Goal: parentStartTime + latency = childStartTime + adjustment
+	return parentStartTime.Add(latency).Sub(childStartTime)
+}
+
+func (a *clockSkewAdjuster) adjustTimestamps(n *node, skew clockSkew) {
+	if skew.delta == 0 {
+		return
+	}
+	if absDuration(skew.delta) > a.maxDelta {
+		if a.maxDelta == 0 {
+			jptrace.AddWarning(n.span, fmt.Sprintf(warningSkewAdjustDisabled, skew.delta))
+			return
+		}
+		jptrace.AddWarning(n.span, fmt.Sprintf(warningMaxDeltaExceeded, a.maxDelta, skew.delta))
+		return
+	}
+	n.span.SetStartTimestamp(pcommon.NewTimestampFromTime(n.span.StartTimestamp().AsTime().Add(skew.delta)))
+	jptrace.AddWarning(n.span, fmt.Sprintf("This span's timestamps were adjusted by %v", skew.delta))
```

Comment:
How does %v render duration? Will it add units?

Reply:
@yurishkuro Yeah it does - here's an example

```
This span's timestamps were adjusted by 35ms
```